---
title: "Evaluating Gene Set Enrichment Analysis Methods (methylGSA & ebGSEA) for Whole Genome Bisulfite Sequencing Data"
author: "Maya Arvanitis, McGill University"
date: "Fall 2024"
output:
  html_document: default
  pdf_document: default
---

---
output:
  html_document: default
  pdf_document: default
---

---
## 1. Installations
```{r}
#options for Rmd. formatting
options(repos = c(CRAN = "https://cran.rstudio.com/"))
```

```{r, results="hide", message=FALSE, warning=FALSE}
#installation
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ChAMPdata")
BiocManager::install("ChAMP")

#Installing dependencies
install.packages("remotes")
remotes::install_github("YuLab-SMU/ggtree")
BiocManager::install(c("enrichplot", "clusterProfiler", "methylGSA"), dependencies = TRUE)
BiocManager::install("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
```

```{r, results="hide", message=FALSE, warning=FALSE}
library(ChAMP)
tools::package_dependencies("methylGSA", recursive = TRUE)
library(methylGSA)
```

## 2. Running on Toy Data

#### methylGSA
----------------------------------------
First, for visualization / understanding of parameters, we will run the two methods on toy data given in documentation:

```{r}
vignette("methylGSA-vignette", package = "methylGSA")
data(package = "methylGSA")
data(cpgtoy)
head(cpg.pval, 20)
```
We will run the two main functions from methylGSA:
1. methylglm: Implements logistic regression adjusting for number of probes in enrichment analysis. 
2. methylRRA:Enrichment analysis after adjusting multiple p-values of each gene by Robust Rank Aggregation.

###### 1. methylglm
Usage: 
`methylglm(cpg.pval, array.type = "450K", FullAnnot = NULL,group = "all", GS.list = NULL, GS.idtype = "SYMBOL", GS.type = "GO", minsize = 100, maxsize = 500, parallel = FALSE, BPPARAM = bpparam())`

###### 2. methylRRA
Usage: 
`methylRRA(cpg.pval, array.type = "450K", FullAnnot = NULL,group = "all", method = "ORA", sig.cut = 0.05, topDE = NULL,GS.list = NULL, GS.idtype = "SYMBOL", GS.type = "GO",minsize = 100, maxsize = 500)`

```{r}
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
```
Running methylglm:
```{r}
library(clusterProfiler)
getOption("clusterProfiler.download.method")
install.packages("R.utils")
library("R.utils")
R.utils::setOption("clusterProfiler.download.method", "auto")
```

Note that original documentation uses KEGG sets, but here used GO due to network issues which is why we get low pvalues.
```{r}
res1 = methylglm(cpg.pval = cpg.pval, minsize = 200, 
                 maxsize = 500, GS.type = "GO")
head(res1, 15)
```
We receive a data frame, ranked p values (lowest to highest). With the following columns:
- ID: Gene set ID
- Description: Gene set description
- Size: Number of genes in gene set
- pvalue: p-value in logistic regression
- padg: Adjusted p-value 

Running methylRRA:
This will aggregate several ranked gene lists into a single gene lists. 
```{r}
res2 = methylRRA(cpg.pval = cpg.pval, method = "ORA", 
                   minsize = 200, maxsize = 210)
head(res2, 15)
```

Where we have outputted the following columns:
- ID: Gene set ID
- Description: Gene set description
- Count: Number of significant genes in the gene set
- overlap: Names of significant genes in the gene set
- Size: number of genes in gene set
- pvalue: p-value in ORA
- padj: Adjusted p-value

Running methylgometh: which will apply the GSEAPreranked appraoched using method = "GSEA" in methylRRA
```{r}
res3 = methylRRA(cpg.pval = cpg.pval, method = "GSEA", minsize = 200, maxsize = 210)

head(res3, 10)
```

Where we have outputted the following columns: 
- ID: Gene set ID
- Description: Gene set description
- Size: Number of genes in gene set
- enrichmentScore: Enrichment Score (calcualted to reflect the degree which a set S is overrepresented at the extremes (top of bottom) of the entire list L. See [1] for more details on this + p-value calculation)
- NES: Normalized enrichment score
- pvalue: p-value in GSEA 
- padj: Adjusted p-value

Visaulizing results of methylglm:
```{r}
barplot(res1, num=8, colorby= "pvalue")
```

#### ebGSEA
-------------------------------
Usage: 
`champ.ebGSEA(beta=myNorm, pheno=myLoad$pd$Sample_Group, minN=5, adjPval=0.05, arraytype="450K", cores=1)`


```{r}

#  myLoad <- champ.load(directory=system.file("extdata",package="ChAMPdata"))
#         myNorm <- champ.norm()
#         myGSEA.ebGSEA <- champ.ebGSEA(beta=myNorm,pheno=myLoad$pd$Sample_Group,arraytype="450K")
#         
#         
# sgt.m <-doGT(phenoSMK.v,dataSMK.m,array="450k",ncores=4)
# champ.ebGSEA(beta=myNorm, pheno=myLoad$pd$Sample_Group, minN=5, adjPval=0.05, arraytype="450K", cores=1)

```


[1] Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles. Proceedings of the National Academy of Sciences 102 (43). National Acad Sciences: 15545â€“50.


