---
title: "Evaluating Gene Set Enrichment Analysis Methods (methylGSA & ebGSEA) for Whole Genome Bisulfite Sequencing Data"
author: "Maya Arvanitis, McGill University"
date: "Fall 2024"
output:
  html_document: default
  pdf_document: default
---

---
output:
  html_document: default
  pdf_document: default
---

---
## 1. Installations
```{r}
#options for Rmd. formatting
options(repos = c(CRAN = "https://cran.rstudio.com/"))
```

```{r, results="hide", message=FALSE, warning=FALSE}
#installation
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ChAMPdata")
BiocManager::install("ChAMP")

#Installing dependencies
install.packages("remotes")
remotes::install_github("YuLab-SMU/ggtree")
BiocManager::install(c("enrichplot", "clusterProfiler", "methylGSA"), dependencies = TRUE)
BiocManager::install("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
```

```{r, results="hide", message=FALSE, warning=FALSE}
library(ChAMP)
tools::package_dependencies("methylGSA", recursive = TRUE)
library(methylGSA)
```

## 2. Running on Toy Data

#### methylGSA
----------------------------------------
First, for visualization / understanding of parameters, we will run the two methods on toy data given in documentation:

```{r}
vignette("methylGSA-vignette", package = "methylGSA")
data(package = "methylGSA")
data(cpgtoy)
head(cpg.pval, 20)
```
We will run the two main functions from methylGSA:
1. methylglm: Implements logistic regression adjusting for number of probes in enrichment analysis. 
2. methylRRA:Enrichment analysis after adjusting multiple p-values of each gene by Robust Rank Aggregation.

###### 1. methylglm
Usage: 
`methylglm(cpg.pval, array.type = "450K", FullAnnot = NULL,group = "all", GS.list = NULL, GS.idtype = "SYMBOL", GS.type = "GO", minsize = 100, maxsize = 500, parallel = FALSE, BPPARAM = bpparam())`

###### 2. methylRRA
Usage: 
`methylRRA(cpg.pval, array.type = "450K", FullAnnot = NULL,group = "all", method = "ORA", sig.cut = 0.05, topDE = NULL,GS.list = NULL, GS.idtype = "SYMBOL", GS.type = "GO",minsize = 100, maxsize = 500)`

```{r}
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
```
Running methylglm:
```{r}
library(clusterProfiler)
getOption("clusterProfiler.download.method")
install.packages("R.utils")
library("R.utils")
R.utils::setOption("clusterProfiler.download.method", "auto")
```

Note that original documentation uses KEGG sets, but here used GO due to network issues which is why we get low pvalues.
```{r}
res1 = methylglm(cpg.pval = cpg.pval, minsize = 200, 
                 maxsize = 500, GS.type = "GO")
head(res1, 15)
```
We receive a data frame, ranked p values (lowest to highest). With the following columns:
- ID: Gene set ID
- Description: Gene set description
- Size: Number of genes in gene set
- pvalue: p-value in logistic regression
- padg: Adjusted p-value 

Running methylRRA:
This will aggregate several ranked gene lists into a single gene lists. 
```{r}
res2 = methylRRA(cpg.pval = cpg.pval, method = "ORA", 
                   minsize = 200, maxsize = 210)
head(res2, 15)
```

Where we have outputted the following columns:
- ID: Gene set ID
- Description: Gene set description
- Count: Number of significant genes in the gene set
- overlap: Names of significant genes in the gene set
- Size: number of genes in gene set
- pvalue: p-value in ORA
- padj: Adjusted p-value

Running methylgometh: which will apply the GSEAPreranked appraoched using method = "GSEA" in methylRRA
```{r}
res3 = methylRRA(cpg.pval = cpg.pval, method = "GSEA", minsize = 200, maxsize = 210)

head(res3, 10)
```

Where we have outputted the following columns: 
- ID: Gene set ID
- Description: Gene set description
- Size: Number of genes in gene set
- enrichmentScore: Enrichment Score (calcualted to reflect the degree which a set S is overrepresented at the extremes (top of bottom) of the entire list L. See [1] for more details on this + p-value calculation)
- NES: Normalized enrichment score
- pvalue: p-value in GSEA 
- padj: Adjusted p-value

Visaulizing results of methylglm:
```{r}
barplot(res1, num=8, colorby= "pvalue")
```

#### ebGSEA
-------------------------------
Usage: 
`champ.ebGSEA(beta=myNorm, pheno=myLoad$pd$Sample_Group, minN=5, adjPval=0.05, arraytype="450K", cores=1)`

ebGSEA uses empirical Bayes Gene Set Enrichment Analysis, which will not rank CpGs. Instead it will rank genes according to the overall level of differential methylation of its CpGs/probes, allowing unbiased and sensitive detection of enriched pathways [2].

We will use the toy data from the ebGSEA vignette. This is a subset of a HM450k buccal swab dataset with 325 buccal swab samples with smoking-pack-years as phenotype and 7933 CpG probes.

```{r}
install.packages("ebGSEA_0.1.0.tar.gz", repos=NULL, type="source")
```

```{r}
load("./data/ebgseaDATA.rda")
ls();
print("Dimension of the dataset:")
dim(dataSMK.m)
```

To rank the genes according to levels of differential methylation of probes mapping to each gene, use the global test function. Doing this with the doGT function.

Inputs to the doGT function are the following:
- pheno.v: Vector containing phenotype info, matched with columns of the input DNAm data matrix. For this toy dataset: smoking pack years information
- data.m: matrix of DNAm beta values with probes along rows and samples along columns.
- model: regression model for the global test. defaulted to "linear"
- array: array type for the input data. "450k" = IlluminaHumanMethylation450 data and "850k" for IlluminaMethylationEpic data.
- nscores: Number of cores used for parallel running, defaulted to = 4. 

```{r}
library(ebGSEA)
```
```{r}
sgt.m <-doGT(phenoSMK.v,dataSMK.m,array="450k",ncores=4)
head(sgt.m)
```
The ouput of the doGT function is the following matrix: 
- rows label the genes
- ordered by their overall level of methylation 
- last column = the number of CpGs/probes mapping to a given gene. 

After this procedure, we can run the doGSEAwt function. This fucntion does a pathway enrichment analysis in a threshold independent manner. 

It uses the following tests: 
- The Wilcox test (WT) and the Known-Population Median Test (KPMT) 

The following are the input parameters:
- rankEID.m = matrix output from the doGT function with the rows as lavelled genes (the rows are named by Gene IDs)
- ptw.ls = list of vectors consisting of Gene EntrezIDs of genes pathways of interest. 
- nscores = number of cores used for parallel running, defaulted again to = 4
- minN = minimum number of genes for each pathway available in the ranked gene list to conduct GSEA. If less than this val, p-val of this pathway set to 1. Defaulted to = 5
- adjPVth = adjusted p-val threshold to declare a pathway to be signficantly enriched. (p-val derived from Wilcoson rank sum test and then adjusted with BH method, defaulted to 0.05)

```{r}
data("MSigDB-28Feb14-data")
topGSEA.lm <- doGSEAwt(rankEID.m = sgt.m, ptw.ls = listEZ.lv, ncores = 4, minN = 5,adjPVth = 0.05)
```

The ouput of the doGSEAwt function is a list and consists of the following three objects; 
(1) Rank(P): a matrix showing enriched pathways ranked by adjusted Wilcox test P-values
(2) Rank(AUC): a matrix showing enriched pathways ranked by AUC 
(3) Genestat: list of gene symbols in each enrihced pathway. Each object contains the stat and p-value from global test of each gene. 

```{r}
topGSEA.lm[[1]]
```
Note columns outputted are: 
- nREP: # of genes mapped in this pathway and present on array
- AUC: area under curve from Wilcox test
- P(WT): p-value from Known Population Median Test
- adjP: Adjusted P-value for each pathway, using BH method.

Now, we can look at the Genestat list to see the statistic and p-val from global test for each gene in a specified enriched pathway. 

For example, let's look at the genes from the first pathway 
```{r}
head(topGSEA.lm$Genestat[[1]])
```
Then, plot the AUC and adjP for each enriched pathway for visualization purposes.
```{r}
plot(x = topGSEA.lm[[1]][,2], y = -log10(topGSEA.lm[[1]][,5]), xlab ="AUC", ylab = "-log10(adjP)", main = "AUC and adjP for each enriched pathway", pch = 21, bg = "red")
```

ebGSEA can also perform GSEA with Fisher's Exact Test. Here, we would not need to rank genes and could input a set of CpGs. 

We use the selEIDfromSelCpG function to derive list of genes by implementing a binomial test to determine stat signif. of a gene being selected on account of the number of CpGs mapping to the gene. 

Note the folllowing input parameters for the selEIDfromSelCpG function: 
- selCpG.v: vector of user selected CpGs
- allCpG.v: vector of all CpGs in the input DNAm data matrix
- pvth: p-val threshold to infer the number of selected CpGs mapped to a gene is significant or not in a binomial test. Defaulted to 0.3/length(selCpG.v)
- array: array type for the input CpGs. "450k" for IlluminaHumanMethylation450 data and "850k" for IlluminaMethylaltionEPIC data. 

From the ebGSEA documentation, using the complete buccal swab dataset. It is a set of 40626 CpGs that exhibit differential methylation associated with smoking pack-years. 
```{r}
data("SampleCpG")
sigEID.ls <- selEIDfromSelCpG(selCpG.v = sampleCpG.v, allCpG.v = allCpG.v, array = "450k")
```

We find that the group of CpGs are significantly mapped to 255 genes.
```{r}
summary(sigEID.ls)
```
Last, we will apply a final function doGSEAft to do a pathway enrichment analysis with Fisher's Exact Test. 

We have the following input parameters for the doGSEAft function: 
- selEID.v: a vector of selected Entrez Gene ID
- ptw.ls: list of Gene EntrezID in each pathway of interest.
- allEID.v: vector of universal set of Entrez Gene ID which selecting genes from 
- nscores: Number of scores for parallel running, defaulted to 4
- minN: for each pathway, min number of genes (in the ranked gene list) to conduct GSEA. If less than this value, p-val of this pathway set to 1. Defaulted to 5.
- adjPVth: adjusted p-val threshold to infer if pathway is significantly enriched or not. P value is calculated from Wilcoxon rank sum test, adjusted with BH method. Defaulted to 0.05

```{r}
topGSEAft.lm <- doGSEAft(selEID.v = sigEID.ls$selEID, ptw.ls = listEZ.lv, allEID.v = names(mapEIDto450k.lv), ncores = 1, adjPVth = 0.05)
```

Ouput of the topGSEAft.lm function is the following: 
- Rank(P): matrix with enriched pathways ranked by adjusted Fisher's Exact Test P-values. 
    - nREP = number of genes in pathway 
    - nOVL = number of selected genes in the pathway 
    - OR = odds ratio of Fisher's Exact Test
    - P = p-value of Fisher's Exact Test
    - adjP = adjusted P-value of Fisher's Exact Test (method= "BH")
    - Genes = all selected genes in the pathway
- Rank(OR): matrix showing enriched pathways but ranked by odds ratio. Same columns defined as in above for Rank(P) 

```{r}
summary(topGSEAft.lm)
```

```{r}
head(topGSEAft.lm$`Rank(P)`)
```

From this, we can see that lung cancer appears at the top telling us that lung cancer is strongly correlated to smoking. Our expected results from this dataset. 

For visualization purposes, odds ratio and adjPval can be plotted for each enriched pathway.

```{r}
plot(x = log2(as.numeric(topGSEAft.lm[[1]][,3])), y = -log10(as.numeric(topGSEAft.lm[[2]][,5])), xlab ="log2(OR)", ylab = "-log10(adjP)", main = "OR and adjP for each enriched pathway", pch = 21, bg = "red")
```
Now, we have conducted both methods for toy datasets and are ready to use this on our own data to draw conclusions. 

[1] Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles. Proceedings of the National Academy of Sciences 102 (43). National Acad Sciences: 15545–50.

[2] Dong Dong, Yuan Tian, Shu-Cheng Zheng, and Andrew E. Teschendorff. ebgsea: An improved gene set enrichment analysis method for epigenome-wide-association studies. Bioinformatics, 35(18):3514–3516, Sep 2019.

